nixos/opensnitch: Add support for EPBF process monitor

onny opened the pull request:
###### Description of changes

Adding support for EPBF process monitor. As described [here](https://github.com/evilsocket/opensnitch/wiki/monitor-method-ebpf) in the manual, this is the default monitor mode for OpenSnitch because it is more efficent and secure.

Nftables usage and epbf monitor mode is now default as defined in [upstream configuration](https://github.com/evilsocket/opensnitch/blob/master/daemon/default-config.json). The old monitor method can be configured like this:
```
services.opensnitch = {
  enable = true;
  settings.ProcMonitorMethod = "proc";
};
```

Optionally depends on upstream ability to configure the kernel module path https://github.com/evilsocket/opensnitch/issues/928

Sucessfull build depends on opensnitch version update https://github.com/NixOS/nixpkgs/pull/246373

Work was done by @Slime90 which I'll add as a co-author to this PR.

Fixes https://github.com/NixOS/nixpkgs/issues/227294

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [x] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

onny committed:
nixos/opensnitch: Add support for EPBF process monitor

Co-authored-by: Slime90

Slime90 commented:
@onny looks good, I did some build tests last night, can you add a conditional for kernel version to be at least 5.10?

`opensnitch-ebpf = if lib.versionAtLeast kernel.version "5.10” then callPackage ../os-specific/linux/opensnitch-ebpf.nix { } else null; `

onny commented:
@Slime90 Added the changes. I'm going to test it on my local setup :)

Slime90 commented:
@onny i checked the repl to see which kernel packages 22.11 currently has, and built the ebpf module backward until it failed. Everything from 5.10 and up built fine.

I’m not sure I see any new config to install the module though in the service module, I think that should be achievable.

onny commented:
If we target master after 23.05 release we might drop this kernel version condition?

SuperSandro2000 commented on pkgs/os-specific/linux/opensnitch-ebpf/default.nix:
```diff
@@ -0,0 +1,95 @@
+  buildPhase = ''
+    runHook preBuild
+    make headers_install
+    cd samples/bpf && make
```
```suggestion
    cd samples/bpf
    make
```

SuperSandro2000 commented on pkgs/os-specific/linux/opensnitch-ebpf/default.nix:
```diff
@@ -0,0 +1,95 @@
+    runHook preConfigure
+    ebpf="opensnitch-source/ebpf_prog"
+    cp $ebpf/opensnitch*.c $ebpf/common*.h $ebpf/Makefile ${kernel.name}/samples/bpf
+    patchShebangs .
```
Can we be more specific?

SuperSandro2000 commented on pkgs/os-specific/linux/opensnitch-ebpf/default.nix:
```diff
@@ -0,0 +1,95 @@
+  #https://aur.archlinux.org/cgit/aur.git/?h=opensnitch-ebpf-module-git
+  #docs in the repo for patching/building aren't 100 percent up to date
+
+  patchPhase = ''
```
```suggestion
  postPatch = ''
```

SuperSandro2000 commented on pkgs/os-specific/linux/opensnitch-ebpf/default.nix:
```diff
@@ -0,0 +1,95 @@
+    elfutils
+    flex
+    bison
+    libressl
```
Isn't this a library?

SuperSandro2000 commented on pkgs/os-specific/linux/opensnitch-ebpf/default.nix:
```diff
@@ -0,0 +1,95 @@
+
+  srcs = [
+    kernel.src
+    (pkgs.opensnitch.src.overrideAttrs (old: { name = "opensnitch-source"; }))
```
Can't we use a mv, cp or symlink instead?

Slime90 commented:
@onny thanks for updating the PR to include the minimum kernel version. I added a few review comments to address some of @SuperSandro2000 ‘s concerns, and a few of my own. I think it’s pretty close, just needs some cleanup and a tweak so that it builds both for all kernel types, the current pr has an issue building both zen and non zen due to folder unpack name

cole-h commented:
@ofborg eval

Slime90 commented:
@onny I know you are probably very busy, just checking to see if we can push this over the finish line now that 23.05 release is past.

Slime90 commented:
@onny I can take over but unless I misunderstand something, I would need to merge changes from my dev branch into yours, before they appear here would I not?

MinerSebas commented on pkgs/os-specific/linux/opensnitch-ebpf/default.nix:
```diff
@@ -0,0 +1,94 @@
+
+stdenv.mkDerivation rec {
+  pname = "opensnitch_ebpf";
+  version = "1.5.2-${kernel.version}";
```
```suggestion
  version = "${pkgs.opensnitch.version}-${kernel.version}";
```

onny commented:
Updated the PR and simplified it a lot. Depends on Opensnitch version update https://github.com/NixOS/nixpkgs/pull/246373

Going to test it before it's ready for review

Slime90 commented:
Wow, this PR looks much better now, the changes have been very instructive, I have learned alot during this whole process.

One worrying thing from the upstream docs:

`Since v1.6.0, opensnitchd expects to find the opensnitch*.o modules under:
 /usr/local/lib/opensnitchd/ebpf/
 /usr/lib/opensnitchd/ebpf/
 /etc/opensnitchd/ # deprecated, only on < v1.5.x
`

If the new version does not read from `/etc` I am not sure what the best path forward is

onny commented:
> Wow, this PR looks much better now, the changes have been very instructive, I have learned alot during this whole process.
> 
> One worrying thing from the upstream docs:
> 
> `Since v1.6.0, opensnitchd expects to find the opensnitch*.o modules under: /usr/local/lib/opensnitchd/ebpf/ /usr/lib/opensnitchd/ebpf/ /etc/opensnitchd/ # deprecated, only on < v1.5.x `
> 
> If the new version does not read from `/etc` I am not sure what the best path forward is

Thanks for noticing. It is not yet deprecated in 1.6.1 so I'll leave it as it is. I updated the PR and got a bit further, now I get this error:

```
[2023-08-03 14:25:21]  ERR  EBPF-DNS: Failed to attach uprobe uretprobe/gethostbyname : cannot write "r:r___nix_store_1x4ijm9r1a88qk7zcmbbfza324gx1aac_glibc_2_37_8_lib_libc_so_6_117ff0_gobpf_858 /nix/store/1x4ijm9r1a88qk7zcmbbfza324gx1aac-glibc-2.37-8/lib/libc.so.6:0x117ff0\n" to uprobe_events: write /sys/kernel/debug/tracing/uprobe_events: invalid argument
[2023-08-03 14:25:21]  WAR  EBPF-DNS: Unable to attach ebpf listener: cannot write "r:r___nix_store_1x4ijm9r1a88qk7zcmbbfza324gx1aac_glibc_2_37_8_lib_libc_so_6_117ff0_gobpf_858 /nix/store/1x4ijm9r1a88qk7zcmbbfza324gx1aac-glibc-2.37-8/lib/libc.so.6:0x117ff0\n" to uprobe_events: write /sys/kernel/debug/tracing/uprobe_events: invalid argument
```

Not sure how to fix this yet

onny commented:
Works for me now, tested it in a VM :)

onny commented:
Updated PR, now only contains commits related to EPBF process monitor support

MinerSebas reviewed (commented):
1. The new system rules do not work, as no `system-fw.json` is present at `/etc/opensnitchd/system-fw.json`. Though I am unsure whether this is deemed important for the nixos use-case.
2. The used firewall in `services.opensnitch.settings.Firewall` should be set according to `networking.nftables.enable` to either `"iptables"` or `"nftables"`.

onny commented:
> 1. The new system rules do not work, as no `system-fw.json` is present at `/etc/opensnitchd/system-fw.json`. Though I am unsure whether this is deemed important for the nixos use-case.

Since the missing system-fw.json file results in errors in the daemon log, I placed the file into the desired directory. For now the file is not configurable.

>     2. The used firewall in `services.opensnitch.settings.Firewall` should be set according to `networking.nftables.enable` to either `"iptables"` or `"nftables"`.

I thought we already moved to nftables by default since switching to iptables-nftables-compat? https://github.com/NixOS/nixpkgs/pull/81172

onny commented:
Tests introduced with this PR ( https://github.com/NixOS/nixpkgs/pull/248011) succeed on with this PR

onny commented:
> > Wow, this PR looks much better now, the changes have been very instructive, I have learned alot during this whole process.
> > One worrying thing from the upstream docs:
> > `Since v1.6.0, opensnitchd expects to find the opensnitch*.o modules under: /usr/local/lib/opensnitchd/ebpf/ /usr/lib/opensnitchd/ebpf/ /etc/opensnitchd/ # deprecated, only on < v1.5.x `
> > If the new version does not read from `/etc` I am not sure what the best path forward is
> 
> Thanks for noticing. It is not yet removed in 1.6.1 so I'll leave it as it is. I updated the PR and got a bit further, now I get this error:
> 
> ```
> [2023-08-03 14:25:21]  ERR  EBPF-DNS: Failed to attach uprobe uretprobe/gethostbyname : cannot write "r:r___nix_store_1x4ijm9r1a88qk7zcmbbfza324gx1aac_glibc_2_37_8_lib_libc_so_6_117ff0_gobpf_858 /nix/store/1x4ijm9r1a88qk7zcmbbfza324gx1aac-glibc-2.37-8/lib/libc.so.6:0x117ff0\n" to uprobe_events: write /sys/kernel/debug/tracing/uprobe_events: invalid argument
> [2023-08-03 14:25:21]  WAR  EBPF-DNS: Unable to attach ebpf listener: cannot write "r:r___nix_store_1x4ijm9r1a88qk7zcmbbfza324gx1aac_glibc_2_37_8_lib_libc_so_6_117ff0_gobpf_858 /nix/store/1x4ijm9r1a88qk7zcmbbfza324gx1aac-glibc-2.37-8/lib/libc.so.6:0x117ff0\n" to uprobe_events: write /sys/kernel/debug/tracing/uprobe_events: invalid argument
> ```
> 
> Not sure how to fix this yet

Created an upstream bug report on this https://github.com/evilsocket/opensnitch/issues/1013

MinerSebas commented:
> > 1. The new system rules do not work, as no `system-fw.json` is present at `/etc/opensnitchd/system-fw.json`. Though I am unsure whether this is deemed important for the nixos use-case.
> 
> Since the missing system-fw.json file results in errors in the daemon log, I placed the file into the desired directory. For now the file is not configurable.
> 

I tested it, but I think this should reverted. Unlike the Application rules, which are each placed in separate files, the system level rules are constrained to a single File.
This means that with the readonly store symlink, you are stuck with the default settings.

Perhaps as a compromise add a copy of the file, if it doesn't exist yet at system activation?
Though that could be done in a follow up PR.

> > ```
> > 2. The used firewall in `services.opensnitch.settings.Firewall` should be set according to `networking.nftables.enable` to either `"iptables"` or `"nftables"`.
> > ```
> 
> I thought we already moved to nftables by default since switching to iptables-nftables-compat? #81172

A casual glance at that PR, suggests that only the backend was switched to `nftables` which is used by a `iptables` frontend. (With the `iptables-legacy` package using `iptables` for Backend and Frontend, and the `nftables` using `nftables` for both).

Having tested with `networking.nftables.enable` set to true and false, opensnitch continues to work without issue.

onny commented:
@MinerSebas okay fixed it :)
